services/google_drive_service.py

```diff
@@ -2,39 +2,58 @@
+from urllib.parse import urlparse, parse_qs
 from googleapiclient.discovery import build
 from googleapiclient.http import MediaIoBaseDownload
 from googleapiclient.errors import HttpError
 from google.oauth2.service_account import Credentials
+from services.ssm_service import SSMService
 
-SERVICE_ACCOUNT_FILE = os.environ['GOOGLE_SERVICE_ACCOUNT_FILE']
 
 from logger import create_log
 
 logger = create_log(__name__)
 
-service_account_info = json.loads(SERVICE_ACCOUNT_FILE)
-scopes = ['https://www.googleapis.com/auth/drive']
-credentials = Credentials.from_service_account_info(service_account_info, scopes=scopes)
-
-drive_service = build('drive', 'v3', credentials=credentials)
-
-def get_drive_file(file_id: str) -> Optional[BytesIO]:
-    request = drive_service.files().get_media(fileId=file_id)
-    file = BytesIO()
-
-    try:
-        downloader = MediaIoBaseDownload(file, request)
-
-        done = False
-        while not done:
-            status, done = downloader.next_chunk()
-
-    except HttpError as error:
-        logger.warning(f"HTTP error occurred when downloading Drive file {file_id}: {error}")
-        return None
-    except Exception as err:
-        logger.exception(f"Error occurred when downloading Drive file {file_id}")
-        return None
-
-    return file
+class GoogleDriveService:
+    def __init__(self):
+        ssm_service = SSMService()
+        if os.environ['ENVIRONMENT'] == 'production':
+            SERVICE_ACCOUNT_FILE = ssm_service.get_parameter('arn:aws:ssm:us-east-1:946183545209:parameter/drb/production/google-drive-service-key')
+        else:
+            SERVICE_ACCOUNT_FILE = ssm_service.get_parameter('arn:aws:ssm:us-east-1:946183545209:parameter/drb/qa/google-drive-service-key')
+        service_account_info = json.loads(SERVICE_ACCOUNT_FILE)
+        scopes = ['https://www.googleapis.com/auth/drive',
+        'https://www.googleapis.com/auth/drive.file',
+        'https://www.googleapis.com/auth/drive.metadata']
+        credentials = Credentials.from_service_account_info(service_account_info, scopes=scopes)
+        self.drive_service = build('drive', 'v3', credentials=credentials)
+
+    def get_drive_file(self, file_id: str) -> Optional[BytesIO]:
+        request = self.drive_service.files().get_media(fileId=file_id)
+        file = BytesIO()
+
+        try:
+            downloader = MediaIoBaseDownload(file, request)
+
+            done = False
+            while not done:
+                status, done = downloader.next_chunk()
+
+        except HttpError as error:
+            logger.warning(f"HTTP error occurred when downloading Drive file {file_id}: {error}")
+            return None
+        except Exception as err:
+            logger.exception(f"Error occurred when downloading Drive file {file_id}")
+            return None
+
+        return file
+
+    def get_file_metadata(self, file_id: str) -> Optional[str]:
+        # supportsAllDrives=True required as of 12/24 despite deprecation
```

Comment:
is this still needed?

Reply:
I wanted to leave this here in case anyone tried to refactor this code and referred to the Google documentation, which is [in some places incorrect](https://developers.google.com/resources/api-libraries/documentation/drive/v3/python/latest/drive_v3.changes.html). It looks like they added warnings to the codebase that the supportsAllDrives would be considered true by default after 2020, and planned to support Team drives by default, but this isn't the case at least for get_media().

Reply:
I lost a lot of time to this before I just tried adding it anyway - I didn't want the same to happen to anyone else.